server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java

```diff
@@ -167,22 +125,25 @@ public static FieldMapper getFieldMapper(MapperService mapperService, String fie
             if (fieldSuffix != null && context.fieldMapper(fieldName + fieldSuffix) != null) {
                 fieldName = fieldName + fieldSuffix;
             }
-            FieldMapper mapper = getFieldMapper(context.getMapperService(), fieldName);
-            if (mapper == null) {
+            MappedFieldType ft = context.fieldMapper(fieldName);
+            if (ft == null) {
                 // Unmapped fields are not ignored
                 fields.put(fieldOrPattern, weight);
                 continue;
             }
-            if (acceptMetadataField == false && mapper instanceof MetadataFieldMapper) {
+            if (acceptMetadataField == false && ft.name().startsWith("_")) {
                 // Ignore metadata fields
                 continue;
             }
-            // Ignore fields that are not in the allowed mapper types. Some
-            // types do not support term queries, and thus we cannot generate
-            // a special query for them.
-            String mappingType = mapper.fieldType().typeName();
-            if (acceptAllTypes == false && ALLOWED_QUERY_MAPPER_TYPES.contains(mappingType) == false) {
-                continue;
+            if (acceptAllTypes == false) {
+                try {
+                    ft.termQuery("", context);
+                } catch (QueryShardException|UnsupportedOperationException e) {
+                    // field type is never searchable with term queries (eg. geo point): ignore
+                    continue;
+                } catch (RuntimeException e) {
+                    // other exceptions are parsing errors or not indexed fields: keep
+                }
```

Comment:
Is there a similar capability-testing concern with field types that support phrase queries?